src/libopensc/card.c

```diff
@@ -397,41 +397,25 @@ int sc_lock(sc_card_t *card)
 		return r;
 	if (card->lock_count == 0) {
 		if (card->reader->ops->lock != NULL) {
-			r = card->reader->ops->lock(card->reader);
-			while (r == SC_ERROR_CARD_RESET || r == SC_ERROR_READER_REATTACHED) {
+			r = card->reader->ops->lock(card->reader, 0);
+			if (r == SC_ERROR_CARD_RESET) {
 				/* invalidate cache */
 				memset(&card->cache, 0, sizeof(card->cache));
 				card->cache.valid = 0;
-				if (was_reset++ > 4) /* TODO retry a few times */
-					break;
-				r = card->reader->ops->lock(card->reader);
 			}
-			if (r == 0)
-				reader_lock_obtained = 1;
 		}
 		if (r == 0)
 			card->cache.valid = 1;
 	}
 	if (r == 0)
 		card->lock_count++;
 
-	if (r == 0 && was_reset > 0) {
-#ifdef ENABLE_SM
-		if (card->sm_ctx.ops.open)
-			card->sm_ctx.ops.open(card);
-#endif
```

Comment:
@frankmorgner:
> sm channel is not reopened on reset anymore

It is, since a reset triggers card reinitialization and
this results in opening of a sm channel.

BTW. Old code had a difference in behavior there:
in init code the sm channel was opened only in
"Transmit" sm mode, but on reset it was opened
unconditionally.